File: engine/runtime/src/main/scala/org/enso/compiler/ModuleCache.scala
package org.enso.compiler

import java.nio.charset.{Charset, StandardCharsets}
import java.util.logging.Level
import buildinfo.Info
import com.oracle.truffle.api.source.Source
import io.circe.generic.JsonCodec
import io.circe.parser._
import io.circe.syntax._

import org.enso.compiler.core.IR
import org.enso.interpreter.runtime.builtin.Builtins
import org.enso.interpreter.runtime.{EnsoContext, Module}

import scala.jdk.OptionConverters._
import scala.util.{Failure, Success, Try}

// TODO Once #1971 is fixed, the logging statements should go back to using our
//  normal templating syntax.

/** Responsible for the management of caches associated with a given module.
  *
  * @param module the module for which the caches are managed.
  */
class ModuleCache(private val module: Module)
    extends Cache[ModuleCache.CachedModule] {
  import org.enso.compiler.ModuleCache.{
    irCacheDataExtension,
    irCacheMetadataExtension,
    CachedModule
  }

  override type M = ModuleCache.Metadata

  override val logLevel: Level = Level.FINEST

  override def stringRepr: String = module.getName.toString

  override def computeDigestFromSource(
    context: EnsoContext
  ): Option[String] = {
    computeDigestOfModuleSources(module.getSource)
  }

  override def computeDigest(
    entry: ModuleCache.CachedModule
  ): Option[String] = {
    computeDigestOfModuleSources(entry.source)
  }

  private def computeDigestOfModuleSources(source: Source): Option[String] = {
    if (source != null) {
      val sourceBytes = if (source.hasBytes) {
        source.getBytes.toByteArray
      } else {
        source.getCharacters.toString.getBytes(StandardCharsets.UTF_8)
      }
      Some(computeDigestFromBytes(sourceBytes))
    } else None
  }

  override def getCacheRoots(context: EnsoContext): Option[Cache.Roots] = {
    if (module != context.getBuiltins.getModule) {
      context.getPackageOf(module.getSourceFile).toScala.map { pkg =>
        val irCacheRoot    = pkg.getIrCacheRootForPackage(Info.ensoVersion)
        val qualName       = module.getName
        val localCacheRoot = irCacheRoot.resolve(qualName.path.mkString("/"))

        val distribution = context.getDistributionManager
        val pathSegments = List(
          pkg.namespace,
          pkg.name,
          pkg.config.version,
          Info.ensoVersion
        ) ++ qualName.path
        val path = distribution.LocallyInstalledDirectories.irCacheDirectory
          .resolve(pathSegments.mkString("/"))
        val globalCacheRoot = context.getTruffleFile(path.toFile)

        Cache.Roots(localCacheRoot, globalCacheRoot)
      }
    } else {
      val distribution = context.getDistributionManager
      val pathSegments = List(
        Builtins.NAMESPACE,
        Builtins.PACKAGE_NAME,
        Info.ensoVersion,
        Info.ensoVersion
      ) ++ module.getName.path
      val path = distribution.LocallyInstalledDirectories.irCacheDirectory
        .resolve(pathSegments.mkString("/"))
      val globalCacheRoot = context.getTruffleFile(path.toFile)

      Some(Cache.Roots(globalCacheRoot, globalCacheRoot))
    }
  }

  override protected def extractObjectToSerialize(
    entry: CachedModule
  ): AnyRef = {
    entry.module
  }

  override protected def validateReadObject(
    obj: AnyRef,
    meta: M
  ): Try[CachedModule] = obj match {
    case ir: IR.Module =>
      Success(
        CachedModule(
          ir,
          Module.CompilationStage.valueOf(meta.compilationStage),
          module.getSource
        )
      )
    case other =>
      Failure(
        new RuntimeException(
          s"unexpected deserialized object of ${other.getClass}"
        )
      )
  }

  override protected def metadataFromBytes(bytes: Array[Byte]): Option[M] = {
    val maybeJsonString = new String(bytes, ModuleCache.metadataCharset)
    decode[ModuleCache.Metadata](maybeJsonString).toOption
  }

  override protected def metadata(
    sourceDigest: String,
    blobDigest: String,
    entry: CachedModule
  ): Array[Byte] = {
    val metadata = ModuleCache.Metadata(
      sourceDigest,
      blobDigest,
      entry.compilationStage.toString
    )
    metadata.asJson.toString().getBytes(ModuleCache.metadataCharset)
  }

  override protected def entryName: String = module.getName.item

  override protected def dataSuffix: String = irCacheDataExtension

  override protected def metadataSuffix: String = irCacheMetadataExtension
}
object ModuleCache {

  val irCacheDataExtension: String     = ".ir"
  val irCacheMetadataExtension: String = ".meta"
  val metadataCharset: Charset         = StandardCharsets.UTF_8

  /** A representation of a module for the cache.
    *
    * @param module the module
    * @param compilationStage the compilation stage of the module at the point
    *                         at which it was provided for serialisation
    * @param source the source of the module
    */
  case class CachedModule(
    module: IR.Module,
    compilationStage: Module.CompilationStage,
    source: Source
  )

  /** Internal storage for the cache metadata to enable easy serialisation and
    * deserialisation.
    *
    * @param sourceHash the hex-encoded SHA3-224 hash of the source file for
    *                   which the cache was created
    * @param blobHash the hex-encoded SHA3-224 hash of the IR blob as it was
    *                 written to disk
    * @param compilationStage the compilation stage at which the IR blob was
    *                         serialised
    */
  @JsonCodec
  case class Metadata(
    sourceHash: String,
    blobHash: String,
    compilationStage: String
  ) extends Cache.Metadata

}

File: engine/runtime/src/main/java/org/enso/compiler/ModuleCache.java
package org.enso.compiler;

import buildinfo.Info;
import com.fasterxml.jackson.annotation.JsonProperty;
import com.fasterxml.jackson.core.JsonProcessingException;
import com.fasterxml.jackson.databind.ObjectMapper;
import com.oracle.truffle.api.TruffleLogger;
import com.oracle.truffle.api.source.Source;
import java.io.ByteArrayInputStream;
import java.io.ByteArrayOutputStream;
import org.apache.commons.lang3.StringUtils;
import org.enso.compiler.core.IR;
import org.enso.interpreter.runtime.EnsoContext;
import org.enso.interpreter.runtime.Module;
import org.enso.interpreter.runtime.builtin.Builtins;
import org.enso.polyglot.CompilationStage;

import java.io.IOException;
import java.io.ObjectInputStream;
import java.io.ObjectOutputStream;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.Optional;
import java.util.UUID;
import java.util.logging.Level;

public final class ModuleCache extends Cache<ModuleCache.CachedModule, ModuleCache.Metadata> {

    private final Module module;

    public ModuleCache(Module module) {
      super(Level.FINEST, module.getName().toString(), true, false);
      this.module = module;
      this.entryName = module.getName().item();
      this.dataSuffix = irCacheDataExtension;
      this.metadataSuffix = irCacheMetadataExtension;
    }

    @Override
    protected byte[] metadata(String sourceDigest, String blobDigest, CachedModule entry) {
        try {
            return objectMapper.writeValueAsBytes(new Metadata(sourceDigest, blobDigest, entry.compilationStage().toString()));
        } catch (JsonProcessingException e) {
            throw new RuntimeException(e);
        }
    }

    @Override
    protected CachedModule deserialize(EnsoContext context, byte[] data, Metadata meta, TruffleLogger logger) throws ClassNotFoundException, IOException, ClassNotFoundException {
        try (var stream = new ObjectInputStream(new ByteArrayInputStream(data))) {
          if (stream.readObject() instanceof IR.Module ir) {
              try {
                  return new CachedModule(ir,CompilationStage.valueOf(meta.compilationStage()), module.getSource());
              } catch (IOException ioe) {
                  throw new ClassNotFoundException(ioe.getMessage());
              }
          } else {
              throw new ClassNotFoundException("Expected IR.Module, got " + data.getClass());
          }
        }
    }

    @Override
    protected Optional<Metadata> metadataFromBytes(byte[] bytes, TruffleLogger logger) {
        var maybeJsonString = new String(bytes, Cache.metadataCharset);
        try {
            return Optional.of(objectMapper.readValue(maybeJsonString, Metadata.class));
        } catch (JsonProcessingException e) {
            logger.log(logLevel, "Failed to deserialize module's metadata.", e);
            return Optional.empty();
        }
    }

    private Optional<String> computeDigestOfModuleSources(Source source) {
        if (source != null) {
            byte[] sourceBytes;
            if (source.hasBytes()) {
                sourceBytes = source.getBytes().toByteArray();
            } else {
                sourceBytes = source.getCharacters().toString().getBytes(metadataCharset);
            }
            return Optional.of(computeDigestFromBytes(sourceBytes));
        } else {
            return Optional.empty();
        }
    }

    @Override
    protected Optional<String> computeDigest(CachedModule entry, TruffleLogger logger) {
        return computeDigestOfModuleSources(entry.source());
    }

    @Override
    protected Optional<String> computeDigestFromSource(EnsoContext context, TruffleLogger logger) {
        try {
            return computeDigestOfModuleSources(module.getSource());
        } catch (IOException e) {
            logger.log(logLevel, "failed to retrieve the source of " + module.getName(), e);
            return Optional.empty();
        }
    }

    @Override
    protected Optional<Cache.Roots> getCacheRoots(EnsoContext context) {
        if (module != context.getBuiltins().getModule()) {
            return context.getPackageOf(module.getSourceFile()).map(pkg -> {
                var irCacheRoot    = pkg.getIrCacheRootForPackage(Info.ensoVersion());
                var qualName       = module.getName();
                var localCacheRoot = irCacheRoot.resolve(qualName.path().mkString("/"));

                var distribution = context.getDistributionManager();
                var pathSegmentsJava = new ArrayList<String>();
                pathSegmentsJava.addAll(Arrays.asList(
                        pkg.namespace(),
                        pkg.module(),
                        pkg.getConfig().version(),
                        Info.ensoVersion()
                ));
                pathSegmentsJava.addAll(qualName.pathAsJava());
                var path = distribution.LocallyInstalledDirectories().irCacheDirectory()
                        .resolve(StringUtils.join(pathSegmentsJava, "/"));
                var globalCacheRoot = context.getTruffleFile(path.toFile());

                return new Cache.Roots(localCacheRoot, globalCacheRoot);
            });
        } else {
            var distribution = context.getDistributionManager();
            var pathSegmentsJava = new ArrayList<String>();
            pathSegmentsJava.addAll(Arrays.asList(
                    Builtins.NAMESPACE,
                    Builtins.PACKAGE_NAME,
                    Info.ensoVersion(),
                    Info.ensoVersion()
            ));
            pathSegmentsJava.addAll(module.getName().pathAsJava());
            var path = distribution.LocallyInstalledDirectories().irCacheDirectory()
                    .resolve(StringUtils.join(pathSegmentsJava, "/"));
            var globalCacheRoot = context.getTruffleFile(path.toFile());

            return Optional.of(new Cache.Roots(globalCacheRoot, globalCacheRoot));
        }
    }

    @Override
    protected byte[] serialize(EnsoContext context, CachedModule entry) throws IOException {
      var byteStream = new ByteArrayOutputStream();
      boolean noUUIDs = false;
      for (var p : context.getPackageRepository().getLoadedPackagesJava()) {
        if ("Standard".equals(p.namespace())) {
          for (var s : p.listSourcesJava()) {
            if (s.file().getPath().equals(entry.source().getPath())) {
              noUUIDs = true;
              break;
            }
          }
        }
      }
      try (var stream = new ObjectOutputStream(byteStream) {
        void filterUUIDs() {
          enableReplaceObject(true);
        }

        @Override
        protected Object replaceObject(Object obj) throws IOException {
          if (obj instanceof UUID) {
            return null;
          }
          return obj;
        }
      }) {
        if (noUUIDs) {
          stream.filterUUIDs();
        }
        stream.writeObject(entry.moduleIR());
      }
      return byteStream.toByteArray();
    }

    // CachedModule is not a record **on purpose**. There appears to be a Frgaal bug leading to invalid compilation error.
    static class CachedModule {
        private final IR.Module _moduleIR;
        private final CompilationStage _compilationStage;
        private final Source _source;
        public CachedModule(IR.Module moduleIR, CompilationStage compilationStage, Source source) {
            this._moduleIR = moduleIR;
            this._compilationStage = compilationStage;
            this._source = source;
        }

        IR.Module moduleIR() {
            return _moduleIR;
        }
        CompilationStage compilationStage() {
            return _compilationStage;
        }
        Source source() {
            return _source;
        }

    }

    public record Metadata(
            @JsonProperty("source_hash") String sourceHash,
            @JsonProperty("blob_hash") String blobHash,
            @JsonProperty("compilation_stage") String compilationStage) implements Cache.Metadata {}

    private final static String irCacheDataExtension = ".ir";

    private final static String irCacheMetadataExtension = ".meta";

    private final static ObjectMapper objectMapper = new ObjectMapper();

}
